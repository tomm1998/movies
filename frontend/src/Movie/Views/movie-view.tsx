import React, { useState } from "react";
import Button from "../../components/MovieButton";
import { MovieFetcher } from "../../renderProps/MovieFetcher";
import List from "@mui/material/List";
import ListItem, { ListItemProps } from "@mui/material/ListItem";
import ListItemIcon from "@mui/material/ListItemIcon";
import ListItemText from "@mui/material/ListItemText";
import Divider from "@mui/material/Divider";
import { AcUnitSharp, Add, Star, StarBorder } from "@mui/icons-material";
import { useNavigate } from "react-router";
import { MovieDetails } from "../../components/MovieDetails";
import { Fab } from "@mui/material";
import { Context } from "../../hooks/SearchBarContext";
import { Filter } from "../../components/Filter";
import { filterByDate } from "../../utils/filterByDate";
import { withFavorite } from "../../utils/withFavorite";

export const MovieView = () => {
  const navigate = useNavigate();
  const [moviesAndFavorites, setMovies] = React.useState([]);
  const queryRef = React.useRef("");
  const [dateFilter, setDateFilter] = useState({
    from: undefined,
    to: undefined,
  });
  return (
    <>
      <Context.Consumer>
        {({ query }) => (
          <div
            style={{
              height: "700px",
              overflow: "auto",
              boxShadow:
                "-60px 0px 100px -90px #000000, 60px 0px 100px -90px #000000",
            }}
          >
            <MovieFetcher all query={query}>
              {(movies) => {
                const moviesWithFavorites = withFavorite(movies);
                if (
                  (!moviesAndFavorites.length && moviesWithFavorites.length) ||
                  query != queryRef.current
                ) {
                  setMovies(moviesWithFavorites);
                  queryRef.current = query;
                }
                return (
                  <div>
                    <Filter
                      onChange={(from, to) => {
                        setDateFilter({ from, to });
                      }}
                    />
                    <List>
                      {filterByDate(
                        moviesAndFavorites,
                        dateFilter.from,
                        dateFilter.to
                      ).map((movie) => {
                        return (
                          <div
                            onClick={() => {
                              navigate(
                                "/movie/" + movie.movieName.split(" ").join("_")
                              );
                            }}
                          >
                            <ListItem button>
                              <ListItemIcon>
                                <AcUnitSharp />
                              </ListItemIcon>
                              <ListItemText primary={movie.movieName}>
                                {movie.title}
                              </ListItemText>
                              <ListItemIcon
                                onClick={(e) => {
                                  e.stopPropagation();
                                  movie.favorite = !movie.favorite;
                                  console.log(moviesAndFavorites);
                                  setMovies([
                                    ...moviesAndFavorites.filter(
                                      (m) => m.title == movie.title
                                    ),
                                    movie,
                                  ]);
                                }}
                              >
                                {movie.favorite ? <Star /> : <StarBorder />}
                              </ListItemIcon>
                            </ListItem>
                            <Divider />
                          </div>
                        );
                      })}
                    </List>
                  </div>
                );
              }}
            </MovieFetcher>
            <Fab
              style={{
                position: "absolute",
                bottom: 20,
                right: 20,
              }}
              color="primary"
              onClick={() => navigate("/add-movie")}
            >
              <Add />
            </Fab>
          </div>
        )}
      </Context.Consumer>
    </>
  );
};
