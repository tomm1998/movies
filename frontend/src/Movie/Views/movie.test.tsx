import { render } from "@testing-library/react";
import React from "react";
import { MovieView } from "./movie-view";

import { describe, expect, it, vi } from "vitest";

vi.mock("../../renderProps/MovieFetcher", () => ({
  MovieFetcher: ({ children }) =>
    children([
      {
        movieName: "Episode VII â€“ The Force Awakens",
        releaseDate: "2021-03-12",
      },
    ]),
}));
describe("Movies", () => {
  it("renders movies", () => {
    const { container } = render(<MovieView />);
    expect(container).toMatchSnapshot();
  });
});
