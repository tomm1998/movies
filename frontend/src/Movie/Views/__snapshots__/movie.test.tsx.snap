// Vitest Snapshot v1

exports[`Movies > renders movies 1`] = `
<div>
  <div
    style="height: 700px; overflow: auto; box-shadow: -60px 0px 100px -90px #000000, 60px 0px 100px -90px #000000;"
  >
    <div>
      <div
        style="display: grid; grid-template-columns: 1fr 1fr;"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              type="date"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-hdw1oc"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              type="date"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-hdw1oc"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <ul
        class="MuiList-root MuiList-padding css-h4y409-MuiList-root"
      >
        <div>
          <div
            class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-padding MuiListItem-button css-bshv44-MuiButtonBase-root-MuiListItem-root"
            role="button"
            tabindex="0"
          >
            <div
              class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="AcUnitSharpIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M22 11h-4.17l3.24-3.24-1.41-1.42L15 11h-2V9l4.66-4.66-1.42-1.41L13 6.17V2h-2v4.17L7.76 2.93 6.34 4.34 11 9v2H9L4.34 6.34 2.93 7.76 6.17 11H2v2h4.17l-3.24 3.24 1.41 1.42L9 13h2v2l-4.66 4.66 1.42 1.41L11 17.83V22h2v-4.17l3.24 3.24 1.42-1.41L13 15v-2h2l4.66 4.66 1.41-1.42L17.83 13H22v-2z"
                />
              </svg>
            </div>
            <div
              class="MuiListItemText-root css-tlelie-MuiListItemText-root"
            >
              <span
                class="MuiTypography-root MuiTypography-body1 MuiListItemText-primary css-10hburv-MuiTypography-root"
              >
                Episode VII – The Force Awakens
              </span>
            </div>
            <div
              class="MuiListItemIcon-root css-cveggr-MuiListItemIcon-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="StarBorderIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="m22 9.24-7.19-.62L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21 12 17.27 18.18 21l-1.63-7.03L22 9.24zM12 15.4l-3.76 2.27 1-4.28-3.32-2.88 4.38-.38L12 6.1l1.71 4.04 4.38.38-3.32 2.88 1 4.28L12 15.4z"
                />
              </svg>
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
          />
        </div>
      </ul>
    </div>
    <button
      class="MuiButtonBase-root MuiFab-root MuiFab-circular MuiFab-sizeLarge MuiFab-primary css-1suf56w-MuiButtonBase-root-MuiFab-root"
      style="position: absolute; bottom: 20px; right: 20px;"
      tabindex="0"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
